File: iOS/MultiScreenPhoto/Model/Gallery.swift
/*

Copyright (c) 2014 Samsung Electronics

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

*/

import UIKit
import AssetsLibrary

/// A Gallery represents an instance of the photos contained in the device photo gallery.

/// Use this class to o retreive the photos from the device photo gallery
class Gallery: NSObject {
    
    //iOS library to retreive the photos from the device
    var assetsLibrary: ALAssetsLibrary! = ALAssetsLibrary()
    /// Array of albums
    var albums: [ALAssetsGroup] = []
    /// Array of number of assets by album
    var numberOfAssetsAtAlbumIndex: [Int] = []
    /// Array of albums expanded state (true, false)
    var isAlbumExpandedAtIndex: [Bool] = []
    
    /// Gallery shared instance used as singleton
    class var sharedInstance: Gallery {
        struct Static {
            static var instance: Gallery?
            static var token: dispatch_once_t = 0
        }
        
        dispatch_once(&Static.token) {
            Static.instance = Gallery()
        }
        
        return Static.instance!
    }
    
    override init() {
        super.init()
        isAlbumExpandedAtIndex.removeAll(keepCapacity: false)
    }
    
    
    /// Retrieve albums from the device photo gallery and save the data into the albums Array, numberOfAssetsAtAlbumIndex Array and isAlbumExpandedAtIndex Array.
    ///
    /// :param: completionHandler The callback handler,  return true or false
    func retrieveAlbums(completionHandler: ((Bool!) -> Void)!){
        
        albums.removeAll(keepCapacity: false)
        numberOfAssetsAtAlbumIndex.removeAll(keepCapacity: false)
        
        /// Temporay array of albums expanded state (true, false)
        var tempIsAlbumExpandedAtIndex: [Bool] = []
        
        /// Enumeration Block
        let enumGroupBlock: ALAssetsLibraryGroupsEnumerationResultsBlock = {(assetsGroup: ALAssetsGroup!, stop: UnsafeMutablePointer<ObjCBool>) -> Void in
            if (assetsGroup != nil) {
                
                if (assetsGroup.numberOfAssets() > 0){
                    
                    /// Set the Camera Roll to the first position
                    if (UInt32(assetsGroup.valueForProperty("ALAssetsGroupPropertyType").intValue) == ALAssetsGroupSavedPhotos){
                        /// set this album not expanded
                        tempIsAlbumExpandedAtIndex.insert(true, atIndex: 0)
                        /// Adding the assetGroup to the album Array
                        self.albums.insert(assetsGroup, atIndex: 0)
                        /// Adding the number of photos by album assetGroup
                        self.numberOfAssetsAtAlbumIndex.insert(assetsGroup.numberOfAssets(), atIndex: 0)
                    } else {
                        /// set this album not expanded
                        tempIsAlbumExpandedAtIndex.append(false)
                        /// Adding the assetGroup to the album Array
                        self.albums.append(assetsGroup)
                        /// Adding the number of photos by album assetGroup
                        self.numberOfAssetsAtAlbumIndex.append(assetsGroup.numberOfAssets())
                    }
                }
            } else {
                
                if (self.isAlbumExpandedAtIndex.count != tempIsAlbumExpandedAtIndex.count){
                    self.isAlbumExpandedAtIndex = tempIsAlbumExpandedAtIndex
                }
                
                dispatch_async(dispatch_get_main_queue(),{
                    completionHandler(true)
                })
            }
        }
        
        /// On Fail Block
        let enumFail: ALAssetsLibraryAccessFailureBlock = {(error: NSError!) -> () in
            dispatch_async(dispatch_get_main_queue(),{
                completionHandler(false)
            })
        }
        
        // Enumerate the number of assetGroup
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), {
            self.assetsLibrary.enumerateGroupsWithTypes(0xFFFFFFFF, usingBlock: enumGroupBlock, failureBlock: enumFail)
        })
    }
    
    /// Returns the album name
    /// :param:  album index
    /// :return: the album name
    func albumNameAtIndex(albumIndex: Int) -> String{
        return albums[albumIndex].valueForProperty(ALAssetsGroupPropertyName) as String
    }
    
    
    /// Retrieve an image from the device photo gallery
    ///
    /// :param: album index
    /// :param: image index
    /// :param: UIImageView tag id
    /// :param: isThumbnail true or false
    /// :param: completionHandler The callback handler,  returns UIImage and Info
    func requestImageAtIndex(album: Int, index: Int, containerId:Int, isThumbnail: Bool, completionHandler: ((UIImage!, [NSObject: AnyObject]!, Int, Int) -> Void)!){
        
        
        if (albums.count > album && numberOfAssetsAtAlbumIndex.count > 0 && numberOfAssetsAtAlbumIndex[album] > index){
            
            let assetsGroup:ALAssetsGroup = albums[album]
            
            /// Enumerate asset for a given index
            assetsGroup.enumerateAssetsAtIndexes(NSIndexSet(index: index), options: nil, usingBlock: {
                (asset: ALAsset!, index: Int, stop: UnsafeMutablePointer<ObjCBool>) in
                if (asset != nil) {
                    
                    var image: UIImage
                    
                    if (isThumbnail){
                        // If it is a thumbnail then retrieve a small size photo
                        image = self.thumbnailForAsset(asset)
                    } else {
                        // Retrieve a medium size photo
                        image = self.imageForAsset(asset)
                    }
                    
                    dispatch_async(dispatch_get_main_queue(),{
                        return completionHandler(image, nil, index, containerId)
                    })
                }
            })
        } else {
            dispatch_async(dispatch_get_main_queue(),{
                return completionHandler(nil, nil, index, containerId)
            })
        }
        
    }
    
    /// Retrieve a medium size photo
    ///
    /// :param: ALAsset
    /// :return: UIImage to be displayed
    func imageForAsset(asset: ALAsset!) -> UIImage{
        var cgImage: CGImageRef
        if (asset.defaultRepresentation() != nil){
            var assetRep: ALAssetRepresentation = asset.defaultRepresentation()
            cgImage = assetRep.fullScreenImage().takeUnretainedValue()
            return UIImage(CGImage: cgImage)!
        }
        return thumbnailForAsset(asset)
    }
    
    
    /// Retrieve a small size photo
    ///
    /// :param: ALAsset
    /// :return: UIImage to be displayed
    func thumbnailForAsset(asset: ALAsset!) -> UIImage{
        var cgImage: CGImageRef
        if (asset.thumbnail() != nil){
            cgImage = asset.aspectRatioThumbnail().takeUnretainedValue()
            return UIImage(CGImage: cgImage)!
        }
        return UIImage(named: "placeholder-image")!
    }
    
}

File: android/MultiscreenPhoto/app/src/main/java/com/samsung/appsamples/photos/model/Gallery.java
/*******************************************************************************
 * Copyright (c) 2015 Samsung Electronics
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *******************************************************************************/


package com.samsung.appsamples.photos.model;

import android.content.ContentResolver;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.provider.MediaStore;

import com.samsung.appsamples.photos.App;

import java.io.File;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.Executors;
import java.util.concurrent.RejectedExecutionException;

/**
 * Created by Bin Liu on 3/10/2015.
 */

public class Gallery extends ArrayList<Gallery.Bucket> {


    /**
     * Define the thumbnail type to be displayed.
     */
    public static final int THUMB_TYPE = MediaStore.Images.Thumbnails.MINI_KIND;

    /**
     * The photo bucket class.
     */
    public static class Bucket extends CopyOnWriteArrayList<Photo> {
        /**
         * If false, do not display the photos.
         */
        private boolean mVisible = false;
        private int mId;
        private String mName;

        public Bucket(int id, String name) {
            this.setId(id);
            this.setName(name);
            mVisible = false;
        }

        /**
         * Set the visibility of the bucket.
         *
         * @param value the visibility value.
         */
        public void setVisibility(boolean value) {
            mVisible = value;
        }

        /**
         * Check if the bucket is visible.
         *
         * @return true if it is visible otherwise false.
         */
        public boolean isVisible() {
            return mVisible;
        }

        @Override
        public boolean equals(Object o) {
            if (!(o instanceof Bucket) || o == null) {
                return false;
            }

            Bucket b = (Bucket) o;
            return b.getId() == this.getId();
        }

        /**
         * Get bucket id.
         */
        public int getId() {
            return mId;
        }

        /**
         * Set the bucket id.
         *
         * @param id the bucket id.
         */
        public void setId(int id) {
            this.mId = id;
        }

        /**
         * Get bucket name.
         */
        public String getName() {
            return mName;
        }

        /**
         * Set bucket name.
         *
         * @param name the name of bucket.
         */
        public void setName(String name) {
            this.mName = name;
        }
    }

    /**
     * Return bucket by id.
     * @param bucketId the bucket id.
     * @return the bucket found.
     */
    public Bucket getBucketById(int bucketId) {
        Iterator<Bucket> ite = iterator();
        while (ite.hasNext()) {
            Bucket bucket = ite.next();
            if (bucketId == bucket.getId()) {
                return bucket;
            }
        }
        return null;
    }

    /**
     * Get thumbnail path by given image id.
     *
     * @param photo The photo object which need to load thumbnail information.
     */
    public static void loadThumbnailInfo(final Photo photo) {
        if (photo == null) {
            return;
        }

        final ContentResolver resolver = App.getInstance().getContentResolver();
        Cursor cursor = MediaStore.Images.Thumbnails.queryMiniThumbnail(
                resolver, photo.getImageID(),
                THUMB_TYPE,
                null);
        if (cursor != null && cursor.getCount() > 0) {
            cursor.moveToFirst();
            String thumbPath = cursor.getString(cursor.getColumnIndex(MediaStore.Images.Thumbnails.DATA));
            photo.setThumbPath(thumbPath);

            int thumbId = cursor.getInt(cursor.getColumnIndex(MediaStore.Images.Thumbnails._ID));
            photo.setThumbId(thumbId);

            File file = new File(thumbPath);
            if (!file.exists()) {
                forceCreateThumbnailAsync(resolver, photo.getImageID());
            }
        } else {
            forceCreateThumbnailAsync(resolver, photo.getImageID());
        }
        cursor.close();
    }


    /**
     * Force to create the thumbnail if it does not exits.
     *
     * @param resolver the content resolver.
     * @param id       the image id.
     */
    private static void forceCreateThumbnailAsync(final ContentResolver resolver, final int id) {
        if (App.getInstance().getCreateThumbnailsService().isTerminated()) {
            App.getInstance().setCreateThumbnailsService(Executors.newFixedThreadPool(10));
        }

        try {
            App.getInstance().getCreateThumbnailsService().execute(new Runnable() {
                @Override
                public void run() {
                    Bitmap bitmap = null;
                    try {
                        bitmap = forceCreateThumbnail(resolver, id);
                    } catch (Exception e) {
                    } finally {
                        if (bitmap != null && !bitmap.isRecycled()) {
                            bitmap.recycle();
                        }
                    }
                }
            });
        } catch (RejectedExecutionException ree) {
        }


    }

    /**
     * Create thumbnail if it does not exist or return it.
     *
     * @param resolver the content resolver.
     * @param id       the image id.
     * @return the Bitmap of the created thumbnail
     */
    private static Bitmap forceCreateThumbnail(final ContentResolver resolver, final int id) {
        Bitmap bitmap = null;
        try {
            bitmap = MediaStore.Images.Thumbnails.getThumbnail(resolver,
                    id, THUMB_TYPE,
                    App.getInstance().getOptionsLoadingThumbnail());
        } catch (Exception e) {

        }
        return bitmap;
    }
}
